### PR TITLE
peer: manual port-forward override for UPnP-less networks

### PR DESCRIPTION
**Pull request by myleshorton**
> ⚠️ This is the replacement for #500, which was accidentally merged into `fisk/peer-connection-events-A` and then reverted (branch reset to the pre-merge tip). All review history lives on #500; this PR carries the same head commit (0000000000000000000000000000000000000000) so reviewers can pick up where they left off.

## Summary

Two related changes to the manual-port-forward path that ship together:

1. **Consolidates** the implementation into the `portforward` package. The manual forwarder landed in `peer/peer.go` via #466 to keep that PR's review surface small, but architecturally it belongs alongside the UPnP-based `Forwarder` — same interface, same package. This PR's first commit (`22d1533`) is a net-zero relocation: `peer.manualPortForwarder` → `portforward.ManualForwarder`, `peer.manualPort` → `portforward.ParseManualPort`, tests follow.

2. **Adds** the `PeerManualPortKey` setting so the Advanced UI in the Share My Connection screen can configure the port without an env var. The env-var path stays for developer / power-user use.

After both commits, the resolution order in `peer.Client.Start`'s `NewForwarder`:

1. `settings.PeerManualPortKey` if non-zero (UI),
2. else `RADIANCE_PEER_EXTERNAL_PORT` if set + parseable (developer override),
3. else fall back to UPnP discovery (existing behavior).

A non-positive or malformed env-var value is logged and the resolution falls through to UPnP discovery rather than registering a non-listening port with lantern-cloud.

## Why two configuration paths

- **Setting (`PeerManualPortKey`)** — for users on networks where UPnP is unavailable (ISP gateways with UPnP off, double-NAT, residential routers with the feature disabled). The Advanced section in the Share My Connection Flutter screen sets it; persists across restarts. Cleared by setting the value to 0.
- **Env var (`RADIANCE_PEER_EXTERNAL_PORT`)** — developer override for testing on macOS / CLI without a UI build. Same parser, same forwarder.

## What's in this PR

**Consolidation (commit `22d1533`):**
- `portforward/manual.go` (new) — `ManualForwarder`, `NewManualForwarder(port uint16)`, `ParseManualPort(s string) (uint16, error)`.
- `portforward/manual_test.go` (new) — `TestParseManualPort` (9 input cases including boundaries + invalid), `TestManualForwarder` (full `portForwarder` contract).
- `peer/peer.go` — drops `manualPortForwarder` + `manualPort()`; `NewForwarder` now calls `portforward.NewManualForwarder` / `portforward.ParseManualPort`. Drops the `strconv` import.
- `peer/peer_test.go` — drops the now-moved tests.
- `ManualForwarder.MapPort` reports `"manual"` as the method tag (was `"manual-env"` when it only served the env path; the suffix is wrong now that both paths feed in).

**Setting integration (commit `0c84c5a`):**
- `common/settings/settings.go` — adds `PeerManualPortKey` (int; 0 = use UPnP).
- `peer/peer.go` — `NewForwarder` resolution order now checks the setting first, then the env var, then falls through.

## `ManualForwarder` contract

- `MapPort` returns a `Mapping` with `External == Internal == configured port` and `Method = "manual"`. No network roundtrip — the router rule is already in place.
- `UnmapPort` is a no-op (user owns the router rule and removes it manually).
- `StartRenewal` is a no-op (manually-configured rules don't carry a UPnP lease).
- `ExternalIP` returns the empty string deliberately. With a manual port forward we have no UPnP gateway to ask for the WAN address, and probing a public-IP service from the client adds a network roundtrip for information lantern-cloud already has — the server observes the peer's source address on the register call and uses that as the canonical external IP when this field is empty.

## How this was sliced

Stacked on radiance #499 (`fisk/peer-connection-events-A`), itself stacked on #466 → #460 → #458. After cascade rebases through the stack, conflicts on `peer/peer.go` resolve cleanly — the consolidation commit moves code that other commits in the chain don't touch.

## Test plan

- [x] `go test -race -count=1 ./peer/... ./portforward/... ./backend/...` clean on the branch.
- [ ] **Manual smoke** (after build):
  - `RADIANCE_PEER_EXTERNAL_PORT=5698 ./lantern-cli …` → `peer.Client` binds 5698, skips UPnP probe, log line `peer client using manual port forward port=5698 source=RADIANCE_PEER_EXTERNAL_PORT`.
  - From Flutter Advanced → manual port = 5698 → toggle SmC on → same outcome with `source=setting`.
  - Empty setting + missing env var → falls back to UPnP discovery (regression check).
  - Malformed env var (`RADIANCE_PEER_EXTERNAL_PORT=abc`) → warning logged, falls through to UPnP rather than failing Start.

## Dependencies

- radiance #499 (`fisk/peer-connection-events-A`) — this PR's base.
- Foundation chain: #458 → #460 → #466.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: portforward/manual.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	"context"
+	"fmt"
+	"strconv"
+)
+
+// ManualForwarder exposes the same Map/Unmap/StartRenewal/ExternalIP
+// surface as Forwarder but does no UPnP work. The user is expected to
+// have configured a port forward on their router by hand (single-port
+// 1:1 NAT — every consumer router exposes port forwarding as a single
+// port number) and supplied the port number out-of-band.
```
[the remark this answers is not in the document]
Comment (myleshorton) [reply portforward/manual.go]: Fixed in 907c9dc. Reworded the doc to describe the implementation requirement instead of making a router-capability claim: 'this implementation reports the same value for both the external and internal port — callers needing distinct external/internal ports should use the UPnP-based Forwarder (which can negotiate them) or build their own portForwarder.'

[the remark this answers is not in the document]
Comment (myleshorton) [reply portforward/manual.go]: Fixed in 907c9dc. Reworded the doc to describe the implementation requirement instead of making a router-capability claim: 'this implementation reports the same value for both the external and internal port — callers needing distinct external/internal ports should use the UPnP-based Forwarder (which can negotiate them) or build their own portForwarder.'

File: peer/peer.go [hunk 2]
```diff
@@ -209,10 +174,52 @@ func NewClient(cfg Config) (*Client, error) {
...
-					"port", p, "env", env.PeerExternalPort.String())
-				return &manualPortForwarder{port: p}, nil
+			// Manual port-forward override. Use case: networks where
+			// UPnP is disabled or unavailable (router has UPnP off for
+			// security, ISP-provided gateways without IGD, networks
+			// behind double-NAT) but the user has manually configured
+			// a port forward on their router. We trust the user's
+			// config — no UPnP roundtrip — and report the configured
+			// port as both the external and internal port (the 1:1
+			// case every consumer router exposes).
```
[the remark this answers is not in the document]
Comment (myleshorton) [reply peer/peer.go]: Fixed in 907c9dc with the same reframing as the manual.go thread. Comment now describes the manual path's behavior: 'report the configured port as both external and internal. (Distinct external/internal ports are out of scope for the manual path; callers needing that should use the UPnP path, which negotiates each leg separately.)'

[the remark this answers is not in the document]
Comment (myleshorton) [reply peer/peer.go]: Fixed in 907c9dc with the same reframing as the manual.go thread. Comment now describes the manual path's behavior: 'report the configured port as both external and internal. (Distinct external/internal ports are out of scope for the manual path; callers needing that should use the UPnP path, which negotiates each leg separately.)'

File: peer/peer.go [hunk 3]
```diff
@@ -209,10 +174,52 @@ func NewClient(cfg Config) (*Client, error) {
...
+			// and values above the port space (70000 → 4464), which
+			// would register a port we don't listen on (or, worse, one
+			// we do listen on for a different service). Out-of-range
+			// values fall through to env-var and then UPnP as if the
+			// setting were unset.
+			if raw := settings.GetInt(settings.PeerManualPortKey); raw != 0 {
+				if raw < 1 || raw > 65535 {
+					slog.Warn("ignoring out-of-range peer_manual_port setting; falling through to env / UPnP",
+						"value", raw)
+				} else {
```
[the remark this answers is not in the document]
Comment (myleshorton) [reply peer/peer.go]: Fixed in 907c9dc. Extracted the manual-port resolution logic from the NewForwarder closure into a private `pickManualForwarder()` function (line-for-line equivalent, just hoisted), then added `TestPickManualForwarder` covering 10 cases — setting precedence over env, setting-only / env-only / neither, setting out-of-range (positive + negative) falling through, env wins when setting is out-of-range, unparseable env falling through, and the low/high boundary values (1 and 65535). The test exercises the public MapPort surface to verify the chosen port without needing UPnP/network access.

[the remark this answers is not in the document]
Comment (myleshorton) [reply peer/peer.go]: Fixed in 907c9dc. Extracted the manual-port resolution logic from the NewForwarder closure into a private `pickManualForwarder()` function (line-for-line equivalent, just hoisted), then added `TestPickManualForwarder` covering 10 cases — setting precedence over env, setting-only / env-only / neither, setting out-of-range (positive + negative) falling through, env wins when setting is out-of-range, unparseable env falling through, and the low/high boundary values (1 and 65535). The test exercises the public MapPort surface to verify the chosen port without needing UPnP/network access.

File: portforward/manual.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+// MapPort reports the configured port as both external and internal. The
+// router-side rule is already in place; nothing to do at the protocol
+// layer.
+func (m *ManualForwarder) MapPort(_ context.Context, _ uint16, _ string) (*Mapping, error) {
+	return &Mapping{
+		ExternalPort: m.port,
+		InternalPort: m.port,
+		Method:       "manual",
+	}, nil
+}
```
[the remark this answers is not in the document]
Comment (myleshorton) [reply portforward/manual.go:66]: Fixed in eb55d0f. `ManualForwarder.MapPort` now sets `Protocol: "TCP"` to match the UPnP-based Forwarder (which hard-codes the same value at `portforward.go:117`), and defensively returns an error when constructed with `port==0` so a caller that bypasses `pickManualForwarder`'s range-check (programmatic use, tests, future code paths) doesn't silently register port 0 with lantern-cloud. `TestManualForwarder` now asserts Protocol; new `TestManualForwarder_RejectsZeroPort` covers the guard.

[the remark this answers is not in the document]
Comment (myleshorton) [reply portforward/manual.go:66]: Fixed in eb55d0f. `ManualForwarder.MapPort` now sets `Protocol: "TCP"` to match the UPnP-based Forwarder (which hard-codes the same value at `portforward.go:117`), and defensively returns an error when constructed with `port==0` so a caller that bypasses `pickManualForwarder`'s range-check (programmatic use, tests, future code paths) doesn't silently register port 0 with lantern-cloud. `TestManualForwarder` now asserts Protocol; new `TestManualForwarder_RejectsZeroPort` covers the guard.

File: peer/peer.go [hunk 5]
```diff
@@ -647,6 +610,53 @@ func ensurePeerOutboundsBypassVPN(options string) (string, error) {
...
+			slog.Info("peer client using manual port forward",
+				"port", port, "source", "setting")
+			return portforward.NewManualForwarder(port)
+		}
+	}
+	if raw := env.GetString(env.PeerExternalPort); raw != "" {
+		port, err := portforward.ParseManualPort(raw)
+		if err != nil {
+			slog.Warn("ignoring invalid "+env.PeerExternalPort.String(),
+				"value", raw, "error", err)
```
[the remark this answers is not in the document]
Comment (myleshorton) [reply peer/peer.go]: Fixed in eb55d0f. Renamed the slog attribute from `error` to `err` to match the convention used elsewhere in peer.go and across the repo (the SigNoz log-aggregation note in CLAUDE.md specifically calls out `log.Error(ctx, msg, err, attrs)` storing the error in the `err` field, so consistency matters for queries).

[the remark this answers is not in the document]
Comment (myleshorton) [reply peer/peer.go]: Fixed in eb55d0f. Renamed the slog attribute from `error` to `err` to match the convention used elsewhere in peer.go and across the repo (the SigNoz log-aggregation note in CLAUDE.md specifically calls out `log.Error(ctx, msg, err, attrs)` storing the error in the `err` field, so consistency matters for queries).

**Comment by teamlantern** [comment]
This PR has been inactive for 30 days.

Waiting on: none

It will be closed automatically in 7 days if no action is taken.

**Comment by myleshorton** [comment]
Superseded by #589, which carries these commits verbatim.

The manual port-forward override for UPnP-less networks — `RADIANCE_PEER_EXTERNAL_PORT` / `settings.PeerManualPortKey`, the extracted manual forwarder, and `ProbeUPnP` — is in #589, along with this branch's phase-granular `StatusEvent` lifecycle, which won over #472's older `Status{}` reset in `Stop()`.

Two pre-existing items from this branch are carried into #589 unchanged rather than quietly fixed, and are called out in its description: `go vet ./peer/...` reports a `lostcancel` finding on `cancelRun` (reproduces on this branch unmodified; not CI-gating, since `lostcancel` is outside `go test`'s default vet subset), and three files needed `gofmt` (applied in #589 as a separate formatting-only commit).

The branch is left intact — nothing is deleted, and this can be reopened if any of the above is wrong.

Why collapsed instead of merged in sequence: these were cut ~2 months ago and sat 81 commits behind `main`. Each reported `MERGEABLE/CLEAN` only because GitHub measures a PR against *its own parent in the stack*, not against `main`; the stack tip actually conflicted in 6 files. Landing them one at a time would have meant 8 sequential rebase-and-merge cycles re-resolving overlapping conflicts.
